contrib/fmhash/fmhash.c

```diff
@@ -0,0 +1,385 @@
+#endif
+static hash_t
+#if defined(__clang__)
+__attribute__((no_sanitize("unsigned-integer-overflow")))
+#endif
+fnv_32(const void* input, size_t len, seed_t seed) {
+    unsigned char *bp = (unsigned char *)input;	/* start of buffer */
+
+    /*
+     * FNV-1 hash each octet in the buffer
+     */
+    size_t i;
+    for (i = 0; i < len; i++) {
+        /* multiply by the 32 bit FNV magic prime mod 2^32 */
+        seed += (seed<<1) + (seed<<4) + (seed<<7) + (seed<<8) + (seed<<24);
+
+        /* xor the bottom with the current octet */
+        seed ^= (seed_t)*bp++;
+    }
+
+    /* return our new hash value */
+    return seed;
+}
+
+
+/*
+ * Modified Bernstein
+ * http://www.eternallyconfuzzled.com/tuts/algorithms/jsw_tut_hashing.aspx
+ */
+#if defined(__clang__)
+#pragma GCC diagnostic ignored "-Wunknown-attributes"
+#endif
+static hash_t
+#if defined(__clang__)
+__attribute__((no_sanitize("unsigned-integer-overflow")))
+#endif
+djb_hash(const void* input, size_t len, seed_t seed) {
+    const char *p = input;
+    hash_t hash = 5381;
+    size_t i;
+    for (i = 0; i < len; i++) {
+        hash = 33 * hash ^ p[i];
+    }
+
+    return hash + seed;
+}
+
+/*Get 32 bit hash for input*/
+static hash_t
+hash32(const void* input, size_t len, seed_t seed) {
+	hash_t xhash = 0;
+	#ifdef USE_HASH_XXHASH
+	    xhash = XXH32(input, len, seed);
+    #else
+	    xhash = fnv_32(input, len, seed);
+    #endif
+    return xhash;
+}
+
+/*Get 64 bit hash for input*/
+static hash_t
+hash64(const void* input, size_t len, seed_t seed) {
+	hash_t xhash = 0;
+	#ifdef USE_HASH_XXHASH
+        xhash = XXH64(input, len, seed);
+    #else
+        xhash = djb_hash(input, len, seed);
+    #endif
+    return xhash;
+}
+
+static rsRetVal
+hash_wrapper2(struct svar *__restrict__ const sourceVal
+		, struct svar *__restrict__ const seedVal, hash_context_t* hcontext) {
+	DEFiRet;
+	int freeHashStr = 0, success = 0;
+	char *hashStr = NULL;
+    seed_t seed = 0;
+    if(seedVal) {
+        seed = var2Number(seedVal, &success);
+        if (!success) {
+            DBGPRINTF("fmhash: hashXX(string, seed) didn't get a valid 'seed' limit, defaulting hash value to 0");
+            iRet = RS_RET_ERR;
+            FINALIZE;
+        }
+    }
+
+    hashStr = (char*)var2CString(sourceVal, &freeHashStr);
+    size_t len = strlen(hashStr);
+    hcontext->xhash = hcontext->hashXX(hashStr, len, seed);
+	DBGPRINTF("fmhash: hashXX generated hash %" PRIu64 " for string(%.*s)"
+					, hcontext->xhash, (int)len, hashStr);
+finalize_it:
+	if (freeHashStr) {
+		free(hashStr);
+	}
+	RETiRet;
+}
+
+static rsRetVal
+hash_wrapper3(struct svar *__restrict__ const sourceVal, struct svar *__restrict__ const modVal
+		, struct svar *__restrict__ const seedVal, hash_context_t* hcontext) {
+
+	DEFiRet;
+    int success = 0;
+    hash_t xhash = 0;
+	hash_t mod = var2Number(modVal, &success);
+	if (! success) {
+		DBGPRINTF("fmhash: hashXXmod(string, mod)/hash64mod(string, mod, seed) didn't"
```

Comment:
just a remark: is this something that should really go to the debug log, or is it something the user should be warned about?